EIPS/eip-4361.md

```diff
@@ -0,0 +1,312 @@
+author: Wayne Chang (@wyc), Gregory Rocco (@obstropolos), Brantly Millegan (@brantlymillegan), Nick Johnson (@Arachnid) 
+discussions-to: https://ethereum-magicians.org/t/eip-4361-sign-in-with-ethereum/7263
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-10-11
+requires: 55, 137, 155, 191, 1271, 1328
+---
+
+## Abstract
+Sign-In with Ethereum describes how Ethereum accounts authenticate with off-chain services by signing a standard message format parameterized by scope, session details, and security mechanisms (e.g., a nonce). The goals of this specification are to provide a self-custodied alternative to centralized identity providers, improve interoperability across off-chain services for Ethereum-based authentication, and provide wallet vendors a consistent machine-readable message format to achieve improved user experiences and consent management.
+
+## Motivation
+When signing in to popular non-blockchain services today, users will typically use identity providers (IdPs) that are centralized entities with ultimate control over users' identifiers, for example, large internet companies and email providers. Incentives are often misaligned between these parties. Sign-In with Ethereum offers a new self-custodial option for users who wish to assume more control and responsibility over their own digital identity.
+
+Already, many services support workflows to authenticate Ethereum accounts using message signing, such as to establish a cookie-based web session which can manage privileged metadata about the authenticating address. This is an opportunity to standardize the sign-in workflow and improve interoperability across existing services, while also providing wallet vendors a reliable method to identify signing requests as Sign-In with Ethereum requests for improved UX.
+
+This work is sponsored by the Ethereum Foundation and Ethereum Name Service (ENS). It is being developed in the open through a series of recorded community calls and public repositories, and its development is informed by over twenty user interviews with a focus on currently-in-production uses, related prior EIPs, and fits within product roadmaps.
+
+## Specification
+Sign-In with Ethereum works as follows:
+
+1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the [EIP-191](https://eips.ethereum.org/EIPS/eip-191) signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.
+2. The signature is then presented to the server, which checks the signature's validity and message content.
+3. Additional fields, including `expiration-time`, `not-before`, `request-id`, `chain-id`, and `resources` may be incorporated as part of authentication for the session.
```

Comment:
Could we consider the flexibility to support additional AuthN claims, for when it makes sense for the application? I can see some value in being able to switch between "wallet profiles" when it makes sense.

For example in a typical B2B App I'd also expect "department", or in another example, like a game, I could own multiple (in-game) characters under my wallet, would I be able to set an app-specific custom claim eg "character_name"?

Reply:
Hi epappas, please see the Resources section to allow this degree of flexibility. We just want to achieve ETH-based authn with this spec, not anything past that, though we are trying to keep that open ended for other implementers. Please feel free to continue the conversation in ETH Magicians if you feel strongly about it. Thanks.